synfig-studio/src/gui/workarearenderer/renderer_brush_overlay.cpp

```diff
@@ -0,0 +1,177 @@
+**	along with Synfig.  If not, see <https://www.gnu.org/licenses/>.
+ **	\endlegal
+ */
+/* ========================================================================= */
+
+/* === H E A D E R S ======================================================= */
+
+#ifdef USING_PCH
+#	include "pch.h"
+#else
+#ifdef HAVE_CONFIG_H
+#	include <config.h>
+#endif
+
+#include "renderer_brush_overlay.h"
+#include <gui/workarea.h>
+#include <synfig/color.h>
+
+#endif
+
+/* === U S I N G =========================================================== */
+
+using namespace synfig;
+using namespace studio;
+
+/* === M A C R O S ========================================================= */
+
+/* === G L O B A L S ======================================================= */
+
+/* === P R O C E D U R E S ================================================= */
+
+/* === M E T H O D S ======================================================= */
+
+Renderer_BrushOverlay::Renderer_BrushOverlay():
+    overlay_enabled(false)
+{
+}
+
+Renderer_BrushOverlay::~Renderer_BrushOverlay()
+{
+}
+
+void
+Renderer_BrushOverlay::set_overlay_surface(const synfig::Surface& surface, const synfig::Rect& rect)
+{
+    overlay_surface = surface;
+    overlay_rect = rect;
+    overlay_enabled = true;
+    if (get_work_area())
+        get_work_area()->queue_draw();
+}
+
+void
+Renderer_BrushOverlay::clear_overlay()
+{
+    overlay_enabled = false;
+    if (get_work_area())
+        get_work_area()->queue_draw();
+}
+
+void
+Renderer_BrushOverlay::enable_overlay(bool enabled)
+{
+    overlay_enabled = enabled;
+    if (get_work_area())
+        get_work_area()->queue_draw();
+}
+
+bool
+Renderer_BrushOverlay::get_enabled_vfunc() const
+{
+    return overlay_enabled && overlay_surface.is_valid();
+}
+
+void
+Renderer_BrushOverlay::render_vfunc(
+    const Glib::RefPtr<Gdk::Window>& drawable,
+    const Gdk::Rectangle& /*expose_area*/)
+{
+    assert(get_work_area());
+    if (!get_work_area() || !overlay_enabled || !overlay_surface.is_valid())
+        return;
+
+    if (!get_work_area()->get_canvas()) return;
+    const RendDesc &rend_desc = get_work_area()->get_canvas()->rend_desc();
+    int canvas_w = get_work_area()->get_w();
+    int canvas_h = get_work_area()->get_h();
+    Vector canvas_tl = rend_desc.get_tl();
+    Vector canvas_br = rend_desc.get_br();
+
+    if (canvas_w <= 0 || canvas_h <= 0
+        || approximate_equal(canvas_tl[0], canvas_br[0])
+        || approximate_equal(canvas_tl[1], canvas_br[1])) return;
+
+    Cairo::RefPtr<Cairo::Context> cr = drawable->create_cairo_context();
+
+    int width = overlay_surface.get_w();
+    int height = overlay_surface.get_h();
+    if (width <= 0 || height <= 0) return;
+
+    Cairo::RefPtr<Cairo::ImageSurface> cairo_surface =
+        Cairo::ImageSurface::create(Cairo::FORMAT_ARGB32, width, height);
+
+    cairo_surface->flush();
+
+    union { int i; char c[4]; } checker = {0x01020304};
+    bool big_endian = checker.c[0] == 1;
```

Comment:
@ice0 @mohamedAdhamc maybe we should follow the algorithm to detect endianess suggested here:
 https://stackoverflow.com/questions/1001307/detecting-endianness-programmatically-in-a-c-program
instead of using `union` tricks.

Perhaps create it as a method in synfig/os.h?

Copy-and-pasted code:
```c++
enum class endianness
{
    little = 0,
    big = 1,
};

inline endianness get_system_endianness()
{
    const int value { 0x01 };
    const void * address { static_cast<const void *>(&value) };
    const unsigned char * least_significant_address { static_cast<const unsigned char *>(address) };

    return (*least_significant_address == 0x01) ? endianness::little : endianness::big;
}
```